store/model.go

```diff
@@ -0,0 +1,276 @@
+	}
+
+	return m
+}
+
+func (m *Model) ReadTxn(f func(txn *Txn) error) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	txn := &Txn{model: m, readonly: true}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Model) WriteTxn(f func(txn *Txn) error) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	txn := &Txn{model: m}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return txn.Commit()
+}
+
+func (m *Model) FindByID(id es.EntityID, v interface{}) error {
+	return m.ReadTxn(func(txn *Txn) error {
+		return txn.FindByID(id, v)
+	})
+}
+
+func (m *Model) Add(v interface{}) error {
+	return m.WriteTxn(func(txn *Txn) error {
+		return txn.Create(v)
+	})
+}
+
+func (m *Model) Delete(id es.EntityID) error {
+	return m.WriteTxn(func(txn *Txn) error {
+		return txn.Delete(id)
+	})
+}
+
+func (m *Model) Save(v interface{}) error {
+	return m.WriteTxn(func(txn *Txn) error {
+		return txn.Save(v)
+	})
+}
+
+func (m *Model) Has(id es.EntityID) (exists bool, err error) {
+	m.ReadTxn(func(txn *Txn) error {
+		exists, err = txn.Has(id)
+		return err
+	})
+	return
+}
+
+type Txn struct {
+	model     *Model
+	discarded bool
+	commited  bool
+	readonly  bool
+
+	actions []es.Action
+}
+
+type SaveOp struct {
+	Before interface{}
+	After  interface{}
+}
+
+func (t *Txn) Create(new interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	id := getEntityID(new)
+	if id == es.EmptyEntityID {
+		id = setNewEntityID(new)
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return errCantCreateExistingInstance
+	}
+
+	a := es.Action{
+		Type:       es.Create,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    new,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Save(updated interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+
+	id := getEntityID(updated)
+	key := t.model.dsKey.ChildString(id.String())
+	beforeBytes, err := t.model.datastore.Get(key)
+	if err == ds.ErrNotFound {
+		return errCantSaveNonExistentInstance
+	}
+	if err != nil {
+		return err
+	}
+
+	before := reflect.New(t.model.valueType.Elem()).Interface()
+	err = json.Unmarshal(beforeBytes, before)
+	if err != nil {
+		return err
+	}
+	a := es.Action{
+		Type:       es.Save,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   before,
+		Current:    updated,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Delete(id es.EntityID) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return ErrNotFound
+	}
+	a := es.Action{
+		Type:       es.Delete,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
```

Comment:
Is it at all useful to keep Previous here, in case we want to 'undo' something, or rollback a transaction? Or is it easier to just leave this up to the underlying store?

Reply:
Oh, that's sounds interesting. I added it to #15 to keep remember that.